elastic/elastic-agent
# Update event logger configuration via Fleet and environment variable (container command)

## What does this PR do?

If Fleet sends an event logging output configuration different than the one that's running, save it to the encrypted store and re-exec the Elastic-Agent to use the new configuration.

This PR adds the ability to receive event logger configuration via Fleet. Previously only the log level was received via Fleet and persisted.

Fleet can store the logging configuration in the policy via the `overrides` option, whenever the Elastic-Agent receives the policy (including at startup), it correctly parses this configuration. This PR enable those values to be used to configure the event logger.

When a policy is received the policy handler compares `agent.logging.event_data.to_stderr` and `agent.logging.event_data.to_files` with it's current values, if the policy contains different values, they're persisted in the disk store, and the Elastic-Agent re-execs. When it re-starts it reads the new values from the persistent store and applies them.


## Note fore reviewers
~~This PR is built on top of https://github.com/elastic/elastic-agent/pull/4909, hence I'm keeping it in draft until https://github.com/elastic/elastic-agent/pull/4909 is merged and I can rebase onto `main`.~~

Because this PR enables changing the log output (from disk to stderr or vice versa) I believe it's better to just re-start the Elastic-Agent instead of trying to do it at runtime. This will help to keep the logs consistent and avoid any possible race condition or the necessity to lock the logger while stopping/starting new outputs.

## Why is it important?

Once https://github.com/elastic/elastic-agent/pull/4549 gets merged the event logger won't be configurable via Fleet, this PR enables their configuration via Fleet.

## Checklist


## Disruptive User Impact



## How to test this PR locally
### Test the container command
1. Get VM and a Stack deployment (ESS is easier) (the deployment needs to be accessible by the VM) 
2. Get [`mock-es`](https://github.com/elastic/mock-es) and run it to return StatusNotAcceptable on all requests: `mock-es -nonindex 100`
3. Add an output to Fleet pointing to mock-es
4. Create a log file with a few lines, anything will do, the content is not important as long as there are at least a few lines
5. Create a Elastic-Agent policy using the output you just created
6. Add the "Custom Logs Integration" pointing to the file you created
7. Export the following environment variables, adjust the values to your environment:
    - `FLEET_ENROLL=1`
    - `FLEET_URL=https://fleet-server.elastic.co:443`
    - `FLEET_ENROLLMENT_TOKEN=c2VjcmV0LXRva2Vu`
    - `EVENTS_TO_STDERR=true`
8. Run the Elastic-Agent container command: `./elastic-agent container`
9. Assert you can see the dropped event logs on stderr:
    ```json
    {
      "log.level": "warn",
      "@timestamp": "2024-06-13T17:17:39.365Z",
      "message": "Cannot index event publisher.Event{Content:beat.Event{Timestamp:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Meta:null, Fields:null, Private:interface {}(nil), TimeSeries:false}, Flags:0x1, Cache:publisher.EventCache{m:mapstr.M(nil)}, EncodedEvent:(*elasticsearch.encodedEvent)(0xc0031fd480)} (status=406): , dropping event!",
      "component": {
        "binary": "filebeat",
        "dataset": "elastic_agent.filebeat",
        "id": "log-mock-es-4ba82ed3_5321_43cc_9ac4_ac2737e02545",
        "type": "log"
      },
      "log": {
        "source": "log-mock-es-4ba82ed3_5321_43cc_9ac4_ac2737e02545"
      },
      "ecs.version": "1.6.0",
      "log.logger": "elasticsearch",
      "log.origin": {
        "file.line": 490,
        "file.name": "elasticsearch/client.go",
        "function": "github.com/elastic/beats/v7/libbeat/outputs/elasticsearch.(*Client).applyItemStatus"
      },
      "service.name": "filebeat",
      "log.type": "event"
    }
    ```

### Fleet-Managed Elastic-Agent
1. Get VM and a Stack deployment (ESS is easier) (the deployment needs to be accessible by the VM) 
2. Get [`mock-es`](https://github.com/elastic/mock-es) and run it to return StatusNotAcceptable on all requests: `mock-es -nonindex 100`
3. Add an output to Fleet pointing to mock-es
4. Create a log file with a few lines, anything will do, the content is not important as long as there are at least a few lines
5. Create a Elastic-Agent policy using the output you just created
6. Install the Elastic-Agent enrolling in the policy you just created
7. Add the new logging settings to the policy using the API/DevTools. You'll need the policy ID and policy name:
    ```
    PUT kbn:/api/fleet/agent_policies/071c91f3-8fce-480e-b625-381d70f98e53
    {
      "name": "aafoobar",
      "namespace": "default",
      "overrides": {
        "agent": {
          "logging": {
            "event_data": {
              "to_stderr": true,
              "to_files": false
            }
          }
        }
      }
    }
    ```
9. Add the "Custom Logs Integration" pointing to the file you created
10. Assert you can see the dropped event logs on stderr. Run `journalctl -lfu elastic-agent.service` and ensure you see logs like this:
    ```json
    {
      "log.level": "warn",
      "@timestamp": "2024-06-13T17:17:39.365Z",
      "message": "Cannot index event publisher.Event{Content:beat.Event{Timestamp:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Meta:null, Fields:null, Private:interface {}(nil), TimeSeries:false}, Flags:0x1, Cache:publisher.EventCache{m:mapstr.M(nil)}, EncodedEvent:(*elasticsearch.encodedEvent)(0xc0031fd480)} (status=406): , dropping event!",
      "component": {
        "binary": "filebeat",
        "dataset": "elastic_agent.filebeat",
        "id": "log-mock-es-4ba82ed3_5321_43cc_9ac4_ac2737e02545",
        "type": "log"
      },
      "log": {
        "source": "log-mock-es-4ba82ed3_5321_43cc_9ac4_ac2737e02545"
      },
      "ecs.version": "1.6.0",
      "log.logger": "elasticsearch",
      "log.origin": {
        "file.line": 490,
        "file.name": "elasticsearch/client.go",
        "function": "github.com/elastic/beats/v7/libbeat/outputs/elasticsearch.(*Client).applyItemStatus"
      },
      "service.name": "filebeat",
      "log.type": "event"
    }
    ```

## Related issues
- Blocked by https://github.com/elastic/elastic-agent/pull/4909

## Questions to ask yourself

- How are we going to support this in production?
- How are we going to measure its adoption?
- How are we going to debug this?
- What are the metrics I should take care of?
- ...



Closes https://github.com/elastic/elastic-agent/issues/4874
